test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -177,6 +182,66 @@ public void initClient() throws IOException {
         assert hasXPack != null;
         assert nodeVersions != null;
     }
+    
+    // Helper class to check warnings in REST responses with sensitivity to versions
+    // used in the target cluster.
+    protected class VersionSensitiveWarningsHandler implements WarningsHandler {
+        Set<String> requiredSameVersionClusterWarnings = new HashSet<>();
+        Set<String> allowedWarnings = new HashSet<>();
+
+        public void current(String... requiredWarnings) {
+            requiredSameVersionClusterWarnings.addAll(Arrays.asList(requiredWarnings));
+            if(clusterHasACurrentVersionNode()) {
+                // as a convenience add all of the current-version warnings to the allowed
+                // warnings because at least one node in the cluster is a current-version node
+                // and so we assume it can return these responses
+                allowedWarnings.addAll(Arrays.asList(requiredWarnings));
+            }
+        }
+
+        public void other(String... allowedWarnings) {            
```

Comment:
Maybe another name. Like, `compatible` maybe? These are warnings that are emitted by compatible versions.